File: pg/FO/ssap-partner-portal-fo-frontend/src/types/Customer.ts
export interface ICustomerListItem {
  id: string
  name: string
  category: string
  salesforceUpdateAt: string
  updateAt: string
}

interface IGetCustomersRequestSortDefine {
  NAME_ASC: string
  NAME_DESC: string
  SALESFORCE_UPDATE_AT_ASC: string
  SALESFORCE_UPDATE_AT_DESC: string
  UPDATE_AT_ASC: string
  UPDATE_AT_DESC: string
  [key: string]: string
}

export const GET_CUSTOMERS_REQUEST_SORT: IGetCustomersRequestSortDefine = {
  NAME_ASC: 'name:asc',
  NAME_DESC: 'name:desc',
  SALESFORCE_UPDATE_AT_ASC: 'salesforceUpdateAt:asc',
  SALESFORCE_UPDATE_AT_DESC: 'salesforceUpdateAt:desc',
  UPDATE_AT_ASC: 'updateAt:asc',
  UPDATE_AT_DESC: 'updateAt:desc',
}

export interface IGetCustomersRequest {
  name: string
  sort: string
  limit: number
  offsetPage: number
}

export interface IGetCustomersResponse {
  offsetPage: number
  total: number
  customers: ICustomerListItem[]
}

interface ISuggestCustomersSortDefine {
  NAME_ASC: string
  [key: string]: string
}

export const SUGGEST_CUSTOMERS_REQUEST_SORT: ISuggestCustomersSortDefine = {
  NAME_ASC: 'name:asc',
}

export interface ISuggestCustomersRequest {
  sort: string
}

export interface ISuggestCustomer {
  id: string
  name: string
}

export interface ICustomerDetail {
  id: string
  name: string
  category: string
  salesforceCustomerId: string
  salesforceUpdateAt: string
  createId: string
  createUserName: string
  createAt: string
  updateId: string
  updateUserName: string
  updateAt: string
  version: number
}

export interface IGetCustomerByIdResponse extends ICustomerDetail {}

File: pg/iret/FO/ssap-partner-portal-fo-backend/app/schemas/customer.py
from datetime import datetime
from typing import List

from app.resources.const import (
    DefaultPageItemCount,
    GetCustomersSortType,
    SuggestCustomersSortType,
)
from app.schemas.base import CustomBaseModel
from fastapi import Query
from pydantic import Field


class GetCustomerByIdResponse(CustomBaseModel):
    """取引先詳細取得レスポンスクラス"""

    id: str = Field(..., title="取引先ID", example="89cbe2ed-f44c-4a1c-9408-c67b0ca2270d")
    name: str = Field(..., title="取引先名", example="ソニーグループ株式会社")
    category: str = Field(None, title="カテゴリ", example="ソニーグループ")
    salesforce_customer_id: str = Field(
        None, title="Salesforce取引先ID", example="fa5c9f96-af22-bbc6-ac4e-cbcb25ad18e7"
    )
    salesforce_update_at: datetime = Field(
        None, title="Salesforce最終更新日時", example="2020-10-23T03:21:39.356+09:00"
    )
    create_id: str = Field(
        ..., title="作成ID", example="fa5c9f96-af22-bbc6-ac4e-cbcb25ad18e7"
    )
    create_at: datetime = Field(
        ..., title="作成日時", example="2020-10-23T03:21:39.356+09:00"
    )
    update_id: str = Field(
        None, title="更新ID", example="fa5c9f96-af22-bbc6-ac4e-cbcb25ad18e7"
    )
    update_user_name: str = Field(None, title="更新ユーザー名", example="ソニー太郎")
    update_at: datetime = Field(
        None, title="更新日時", example="2020-10-23T03:21:39.356+09:00"
    )
    version: int = Field(None, title="ロックキー（楽観ロック制御）", example=1)


class SuggestCustomersQuery(CustomBaseModel):
    """取引先サジェスト用データ取得クエリクラス"""

    sort: SuggestCustomersSortType = Query(..., title="ソート")


class SuggestCustomersResponse(CustomBaseModel):
    """取引先サジェスト用データ取得レスポンスクラス"""

    id: str = Field(..., title="取引先ID", example="89cbe2ed-f44c-4a1c-9408-c67b0ca2270d")
    name: str = Field(..., title="取引先名", example="ソニーグループ株式会社")


class GetCustomersQuery(CustomBaseModel):
    """取引先一覧取得クエリクラス"""

    name: str = Query(None, title="取引先名")
    sort: GetCustomersSortType = Query(GetCustomersSortType.NAME_ASC, title="ソート")
    limit: int = Query(DefaultPageItemCount.limit, title="最大取得件数", example=20)
    offset_page: int = Query(1, title="リストの中で何ページ目を取得するか", example=1)


class CustomerInfoForGetCustomers(CustomBaseModel):
    """GetCustomersResponse.customersのList要素"""

    id: str = Field(..., title="取引先ID", example="89cbe2ed-f44c-4a1c-9408-c67b0ca2270d")
    name: str = Field(..., title="取引先名", example="ソニーグループ株式会社")
    category: str = Field(None, title="カテゴリ", example="ソニーグループ")
    salesforce_update_at: datetime = Field(
        None, title="Salesforce最終更新日時", example="2020-10-23T03:21:39.356+09:00"
    )
    update_at: datetime = Field(
        None, title="更新日時", example="2020-10-23T03:21:39.356+09:00"
    )


class GetCustomersResponse(CustomBaseModel):
    """取引先一覧取得レスポンスクラス"""

    offset_page: int = Field(None, title="リストの総数内で何ページ目か", example=1)
    total: int = Field(..., title="件数", example=10, ge=0)
    customers: List[CustomerInfoForGetCustomers] = Field(..., title="取引先情報")
